registry/auth.go

```diff
@@ -0,0 +1,215 @@
+package registry
+
+import (
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/alibaba/pouch/apis/types"
+
+	"github.com/Sirupsen/logrus"
+)
+
+type challengeClient struct {
+	realm   string
+	service string
+	scope   string
+	httpCli *http.Client
+}
+
+type registryEndpoint struct {
+	version   string
+	url       *url.URL
+	tlsConfig *tls.Config
+}
+
+type token struct {
+	token string
+}
+
+// Auth authenticates the v1/v2 registry with the credentials
+func (client *Client) Auth(config *types.AuthConfig) (token string, err error) {
+	address := config.ServerAddress
+	if address != "" {
+		address = strings.TrimSuffix(config.ServerAddress, "/")
+	}
+
+	endpoints := genregistryEndpoints(address)
+	for _, endpoint := range endpoints {
+		if endpoint.version == "v2" {
+			token, err = loginV2(endpoint.url, config.Username, config.Password)
+			if err == nil {
+				return
+			}
+		} else {
+			token, err = loginV1(endpoint.url, config.Username, config.Password)
+			if err == nil {
+				return
+			}
+		}
+	}
+
+	return "", err
+}
+
+// TODO: add tls support
+func genregistryEndpoints(addr string) (endpoints []registryEndpoint) {
+	// add v2 registry endpoint first.
+	ep := genV2Endpoints(addr)
+	if ep != (registryEndpoint{}) {
+		endpoints = append(endpoints, ep)
+	}
+
+	// TODO: add v1 registry endpoint.
+
+	return endpoints
+}
+
+func genV1Endpoints(addr string) registryEndpoint {
+	return registryEndpoint{}
+}
+
+func genV2Endpoints(addr string) registryEndpoint {
+	if addr == "" {
+		addr = defaultV2Registry
+	}
+	if !strings.HasPrefix(addr, "http") && !strings.HasPrefix(addr, "https") {
+		addr = "https://" + addr
+	}
+	url, err := url.Parse(addr + "/v2")
+	if err != nil {
+		return registryEndpoint{}
+	}
+
+	return registryEndpoint{
+		version: "v2",
+		url:     url,
+	}
+}
+
+func loginV1(url *url.URL, username, password string) (string, error) {
+	logrus.Infof("attempt to login v1 registry %s", url.String())
+	req, err := http.NewRequest("GET", url.String(), nil)
+	if err != nil {
+		return "", err
+	}
+
+	if username != "" || password != "" {
+		req.SetBasicAuth(username, password)
+	}
+
+	v1AuthClient := &http.Client{}
+
+	resp, err := v1AuthClient.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("failed to login v1 registry %s with http status %s", url.String(), http.StatusText(resp.StatusCode))
+	}
+	token := token{}
+	err = json.NewDecoder(resp.Body).Decode(&token)
+	if err != nil {
+		return "", err
+	}
+
+	return token.token, err
+}
+
+func pingV2(url *url.URL) ([]Challenge, error) {
+	req, err := http.NewRequest("GET", url.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
```

Comment:
I think here we need to add `defer resp.Body.Close()`, otherwise it may cause resource leak in pouch daemon. WDYT? @Ace-Tang

Reply:
yes, I miss this, and both are updated.